File: DataStructure/3.LinkedList/Group3/minyeong/1406.cpp
#include <iostream>
#include <string>
#include <list>
using namespace std;

int main(void)
{
    string s;
    cin >> s;
    int m;
    cin >> m;
    //그냥 int형으로 할 시에 시간 복잡도가 너무 큼.  list를 이용해서 할시 시간 복잡도가 o(1)로 해결가능
    list<char>text(s.begin(),s.end());
     //문자열 s를 한 글자씩 list에 저장
    auto cursor = text.end();
    while(m--)
    {
        char input;
        cin >> input;

        if (input == 'L') {
            if (cursor != text.begin())
                cursor--; 
        }
        else if (input == 'D') {
            if (cursor != text.end())
                cursor++;
        }
        else if (input == 'B') {
            if (cursor != text.begin()) {
                cursor--;
                cursor = text.erase(cursor); 
            }
        }
        else if (input == 'P') {
            char PlusInput;
            cin >> PlusInput;
            cursor = text.insert(cursor, PlusInput); 
            cursor++;
        }
    }


    for (char c : text)
    {
        cout << c;
    }
    return 0;
}

File: DataStructure/3.LinkedList/Group3/dooyeon/1406.py
from collections import deque

s = input().strip()
m = int(input())
text = deque(s)
cursor = len(text)

for _ in range(m):
    order = input().split()
    if order[0] == 'L':
        if cursor > 0:
            cursor -= 1
    elif order[0] == 'D':
        if cursor < len(text):
            cursor += 1
    elif order[0] == 'B':
        if cursor > 0:
            cursor -= 1
            text.remove(text[cursor])
    elif order[0] == 'P':
        text.insert(cursor, order[1])
        cursor += 1

print(''.join(text))
